### DISCUSSION TITLE
New Compact View

### DISCUSSION THREAD

**Discussion by Schlossgeist** [transmission/transmission#5306] [Ideas]
I've been working on improving the compact view. Currently, while the compact view displays more torrents, it also hides more information about the torrents compared to the non-compact view. My idea is to give the user more freedom in what and how information is displayed:
![Screenshot_20230327_051530](https://user-images.githubusercontent.com/74201845/227994305-b0a0bc36-c027-41d6-9e0b-5a79773ca6a3.png)

some details:
- about 20 columns can be activated in the menu, including: Priority, Date added, ETA, Up-/Down-Speeds, Tracker Status and more
- sorting torrents based on the column data is possible by clicking the little arrow in the column header section
- columns are resizeable and reorderable
- non-compact view remains completely unchanged

some things I am uncertain about:
- there are about 20 checkboxes for the columns I've put under the _Desktop_-Tab in the preferences window (unsure whether this is the right place). Are there better solutions for this?
- How should the compact view sorting interact with the non-compact view sorting?

**Comment by ckerr** [comment]
Looks pretty cool.

Some quick thoughts:

- Status, Size, Ratio, Queue should be right-justified
- Path has some edge cases that I don't have a one-size-fits-all answer to, e.g. what do we do when the path is really long, or when the user has an Incomplete dir for some torrents but not others.
- Does "Progress" also double for seed ratio when seeding?
- Do sorting by column & sorting by View > Sort play nicely together, e.g. when clicking on a View > Sort mode not available in the table, do all the columns not show up/down carets?

> there are about 20 checkboxes for the columns I've put under the Desktop-Tab in the preferences window (unsure whether this is the right place). Are there better solutions for this?

Hmm, not sure. That _is_ a lot.

- First I would ask if all 20 need to be there. If there's a column that only 1% of people will ever want, then we probably don't need it.
- If the list is still large, maybe this belongs in a Preferences dialog tab.

> How should the compact view sorting interact with the non-compact view sorting?

Following the principle of least surprise, sorting should stay unchanged when toggling between views.

**Reply by Schlossgeist** [reply]
> Status, Size, Ratio, Queue should be right-justified

You're right. After playing around a bit I figured that every column should be right-aligned except for Name, Path and Private.

> Path has some edge cases that I don't have a one-size-fits-all answer to, e.g. what do we do when the path is really long, or when the user has an Incomplete dir for some torrents but not others.

Currently, if the path is too long, it is elided in the middle. That worked for most of my test cases. What does "incomplete" mean in this context?

> Does "Progress" also double for seed ratio when seeding?

Could you explain this a bit more? The progress bar is a remnant from the TorrentDelegateMin and I just cut away everything except the progessbar.

> Do sorting by column & sorting by View > Sort play nicely together, e.g. when clicking on a View > Sort mode not available in the table, do all the columns not show up/down carets?

The sorting is based on the implementation of QTableView, which means that the sorting selected in the compact view will always carry over to the non-compact view.
When clicking on View > Sort mode the sorting is done with the TorrentFilter, e. g. a proxy filter for the model.
These two don't really work well together, so my suggestion would be that we take the sorting resposibility away from the filter proxy and instead only call sortByColumn(column, order) even when the sort mode is selected via View > Sort mode. This has the added benefit that the sorting is correctly applied even if the corresponding column is not active at the time.

These are the 21 columns I identified as the most useful:
        COL_NAME
        COL_PRIVATE
        COL_STATUS
        COL_SIZE
        COL_SEEDING_TIME
        COL_SEEDS
        COL_PEERS
        COL_DOWN_SPEED
        COL_UP_SPEED
        COL_ETA
        COL_RATIO
        COL_DOWNLOADED
        COL_UPLOADED
        COL_PROGRESS
        COL_TRACKER_STATUS
        COL_ADDED_ON
        COL_LAST_ACTIVE
        COL_PATH
        COL_PRIORITY
        COL_QUEUE_POSITION
        COL_SIZE_LEFT

I am also struggling a bit with the correct implementation of saving the active columns. Because of the big number of columns (and therefore individual options) I decided not to add a new pref key for every column but I am instead manipulating a single string that gets written to the settings file and contains the activation state for every column. I am not sure whether this is the best way to do this. With your approval and after all the above issues are sorted out, I would open a PR so you can take a look at the code.

**Comment by ckerr** [comment]
> What does "incomplete" mean in this context?

When torrents are kept in an intermediate directory while they're being downloaded, then moved into their final destination once the torrent is complete.

> The progress bar is a remnant from the TorrentDelegateMin and I just cut away everything except the progessbar.

Sounds good.

> Because of the big number of columns (and therefore individual options) I decided not to add a new pref key for every column but I am instead manipulating a single string that gets written to the settings file and contains the activation state for every column.

That sounds reasonable.

>  I would open a PR so you can take a look at the code.

I'd love to see it!

**Comment by Pentaphon** [comment]
This is amazing and long overdue. This makes compact view infinitely more useful and many of the users will really appreciate this. 

>* How should the compact view sorting interact with the non-compact view sorting?

The sorting should be the same no matter which view you select. 

> COL_UPLOADED

If you need another useful column. SESSION UPLOADED would be great to show how much each torrent uploaded during your current session. It helps with tracking your seeding progress. Not sure if this is viewable.

**Reply by Schlossgeist** [reply]
> SESSION UPLOADED would be great to show how much each torrent uploaded during your current session.

That sounds like it could be useful. @ckerr Does Transmission keep track of this value somewhere? I could not find it and don't know how it can be calculated with exisiting stats.

**Comment by TheExecutionerGH** [comment]
From where i can download the compact view and dark mode?

**Reply by Schlossgeist** [reply]
This feature is still in development and will hopefully be released in one of the coming minor-version updates of Transmission. What dark mode are you refering to? The screenshot from above is the standard dark breeze theme that comes with the KDE desktop environment and is not something specific to Transmission.

**Comment by student13371** [comment]
Transmission is such an amazing BitTorrent client.

It would be wonderful to have columns to display more data.

I understand that developers are extremely busy.

Are there still active plans to add columns to Transmission GUI?
